### PR TITLE
Add the possibility to launch IPFS server / grid --compute / notebook through a single docker-compose file

### PR DESCRIPTION
**Pull request by cthorey**
I tried running the grid this week-end on my laptop and ended up launching everything with docker to not add stuff to my local machine.  It might be redundant #127 #21 .

I ended up basing the worker image on top of (deepo)[https://github.com/ufoym/deepo] which possesses all modern dl frameworks and ipfs on top of the official image. I also added a jupyter server on 8888.

You can easily test yourself

1. Export your IP to an env variable called HOST_IP. This is going to be use by the worker to connect to the IPFS server which run in a different container.
   ```
   export HOST_IP=YOURHOSTIP
   ```

1. Open the docker-compose.compute.yml
2. Replace the name/email to the one you want to use
3. Build

```
docker-compose -f docker-compose.yml build
```
4. Lauch the IPFS server, the worker as well as a notebook server.

```
docker-compose -f docker-compose.yml up
```
5. It is setup to launch the grid by default, just change --compute to whatever else.

I also added gpu version that works with GPU using **nvidia-docker-compose** instead of **docker-compose**.

The CPU version has been tested on
- OSX using Docker version 17.12.0-ce, docker-compose version 1.19.0
- Ubuntu server on gcloud

The GPU version has been tested on 
- Ubuntu server on gcloud

### CODE REVIEW

**Comment by iamtrask** [comment]
This looks awesome! I'm not sure I am familiar enough with docker to make an objective review of this though - gonna need some help from a fellow contributor with more expertise.

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      - "4001:4001"
+      - "8080:8080"
+      - "5001:5001"
+
+  worker:
+    build:
+      context: ./docker
+      dockerfile: ./Dockerfile.worker
+      args:
+        email: 'openmined@openmined.com'
```
Comment (ianlivingstone) [inline docker-compose.yml]: A small thing I thought about as I read through, these could be defined as environment variables with defaults (or not, which causes docker-compose to fail), instead of having them hard coded to `openmined@openmined.com` and `openmined` respectively.

It might be nice in future if we were to build a setup or install script that handled these things.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+version: "2"
```
Comment (ianlivingstone) [inline docker-compose.yml]: Is there a specific reason you chose version 2 instead of 3?

File: docker/Dockerfile.ipfs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+FROM ipfs/go-ipfs
```
Comment (ianlivingstone) [inline docker/Dockerfile.ipfs]: You can override the start command in the docker-compose.yml file via the `command` block, meaning we wouldn't need our own image for it!

https://docs.docker.com/compose/compose-file/#command

File: docker/Dockerfile.worker-gpu [hunk 4]
```diff
@@ -0,0 +1,31 @@
+FROM ufoym/deepo:all-py36
```
Comment (ianlivingstone) [inline docker/Dockerfile.worker-gpu]: I might be mistaken but I can't find a difference between this image and `docker/Dockerfile.worker` that you've provided above. Is there a reason we need two?

File: docker/Dockerfile.worker-gpu [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+# jupyter
+RUN pip install jupyter
+COPY ./run_jupyter.sh /run_jupyter.sh
+COPY ./jupyter_notebook_config.py /root/.jupyter/
+RUN jupyter nbextension enable --py --sys-prefix widgetsnbextension && \
+    python -m ipykernel.kernelspec
+
+
+ENV PASSWORD=tf
```
Comment (ianlivingstone) [inline docker/Dockerfile.worker-gpu]: Is there a reason we're setting a default password for the jupyter server? Would it be easier for user if there wasn't a password set at all? If we do need to set a default password, it'd be nice if it was documented in the README.md.

File: docker/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+numpy
```
Comment (ianlivingstone) [inline docker/requirements.txt]: Can we pull in the requirement.txt file from repository root? It'd be nice to only have one place to document/manage build dependencies.

File: grid/lib/utils.py [hunk 7]
```diff
@@ -54,6 +55,11 @@ def _attempt_ipfs_connection(ipfs_addr,
     try:
         api = ipfsapi.connect(ipfs_addr, port)
         return api
+    except ConnectionError:
```
Comment (ianlivingstone) [inline grid/lib/utils.py]: Perhaps we could use an environment variable to override the default here? For example:

```python
def get_ipfs_api(mode, ipfs_addr='127.0.0.1', port=5001, max_tries=25):
  ipfs_addr = ipfs_addr if 'IPFS_ADDR' not in os.environ else os.environ['IPFS_ADDR']
```

This way we don't have to try to connect twice, we fallback to the ipfs_addr we were provided and we can document it as a part of the log output in `get_ipfs_api` instead of as a fallback inside `_attempt_ipfs_connection`.

File: docker-compose.gpu.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
+version: "2"
```
[the remark this answers is not in the document]
Comment (ianlivingstone) [reply docker-compose.gpu.yml]: It looks like the latest versions of docker-compose support the runtime parameter inside version 3 of `docker-compose.yml`: https://github.com/docker/compose/pull/5405

I'd suggest we look into leveraging the latest nvidia support for docker.

File: docker/Dockerfile.worker [hunk 9]
```diff
@@ -0,0 +1,31 @@
+FROM ufoym/deepo:py36-cpu
```
Comment (ianlivingstone) [inline docker/Dockerfile.worker]: Is there a reason you're using the same image for the notebook and the worker instead of two separate images or a shared base image?

File: docker/Dockerfile.worker-gpu (hunk 5 above)
Comment (johannes69ki) [reply docker/Dockerfile.worker-gpu]: The newer jupyter versions default to using a login prompt with password. In order to get rid off the login screen and passwords, you can add `c.NotebookApp.token = ""` to `/etc/jupyter/jupyter_notebook_config.py`.

File: docker/Dockerfile.worker-gpu (hunk 4 above)
Comment (johannes69ki) [reply docker/Dockerfile.worker-gpu]: IMHO the base image is an ugly piece of a dockerfile. It has a single `RUN` statement to run all the package installation commands, be it `apt-get`, `pip install`, `git clone`... for each and every software artifact that is installed in the container... (see: https://github.com/ufoym/deepo/blob/master/docker/Dockerfile.all-py27) this approach fundamentally neglects the philospohy of layered dockerfiles... thus everytime a `docker build` fails because of any small issue during software installation, everything that has been done before in the build, is thrown away when a build fails... also it never removes unnecessary temporary data like `rm -rf .git` after every `git clone` or `rm -rf .cache` after every `pip install`... thus this dockerfile becomes unnecessarily bloated...

File: docker-compose.gpu.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (johannes69ki) [reply docker-compose.gpu.yml]: We should definetely make use of `nvidia-docker` only. Adding `devices: /dev/nvidia0` to `docker-compose.yml` is a bad practice and reduces portability. From the best practice guide of NVidia (http://docs.nvidia.com/deeplearning/dgx/pdf/Best-Practices-Docker.pdf):

```One approach to solving this complexity when using containers is to have the NVIDIA
drivers installed in the container and have the character devices mapped corresponding
to the NVIDIA GPUs such as /dev/nvidia0. For this to work, the drivers on the host
(the system that is running the container), must match the version of the driver installed
in the container. This approach drastically reduces the portability of the container.

To make things easier for Docker® containers that are built for GPUs, NVIDIA® has
created nvidia-docker. It is and open-source project hosted on GitHub. It is basically
a wrapper around the docker command that takes care of orchestrating the GPU
containers that are needed for your container to run.

Important It is highly recommended you use nvidia-docker when running a Docker
container that uses GPUs.```

File: docker/jupyter_notebook_config.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import os
+
+from IPython.lib import passwd
+
+c.NotebookApp.ip = '*'
```
Comment (jvmancuso) [inline docker/jupyter_notebook_config.py]: Is there a convenient way of allowing the user to run Jupyter remotely?  Would ideally like to spin up clusters in the cloud and connect them to Grid with this container, and I think that would require being able to connect to Jupyter through ssh or through some secure static hostname/IP. Would like to pass a few flags to enable this, default should definitely have the Jupyter running from localhost.

File: docker-compose.gpu.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (cthorey) [reply docker-compose.gpu.yml]: +1 for overiding compose file. I did not know that. +2 for nvidia-docker :) I manage to use and it works as a charm. 

However, I could not get compose to recognize the runtime arguments. The PR seems to have merged in january, I build compose one commit ahead and still not recognised.. If someone else can give it a try .

In the meantime, you can setup nvidia as the default runtime by changing the daemon.yml for docker. I put the instruction in the README.md, it is a bit of a pain, but it works.

File: docker-compose.yml (hunk 1 above)
Comment (cthorey) [reply docker-compose.yml]: I setup a small bash script to handle those

```
./run_grid.sh [-g <gpu>] [-m <grid-mode>] [-a <host_ip>] [-n <name>] [-e <email>]
```

File: docker-compose.yml (hunk 2 above)
Comment (cthorey) [reply docker-compose.yml]: Switch to 3

File: docker/Dockerfile.worker-gpu (hunk 4 above)
Comment (cthorey) [reply docker/Dockerfile.worker-gpu]: I remove the build on top of  deepo. As recommended by @ianlivingstone, i build the main image on top of nvidia/cuda then add python3.6 and run pip to install other requirements. I did not use the already provided ubuntu installation script are they are installing ipfs which will be useless here.

I still need two images cpu/gpu because installing tensorflow using pip install tensorflow does not allow tensorflow to actually detect the gpus. You can check by running 

```
import tensorflow as tf
sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
```
So we need one image, where tensorflow is installed for cpu usage and one in tensorflow-gpu for those who want to use gpus. Ideally, compose does not build anything and just use **openmined/grid** and **openmined/grid:gpu** where the second  is built using a Dockerfile that looks like 

```
FROM openmined/grid

RUN pip install tensorflow-gpu
```

Using compose, you cannot really build one Dockerfile on top of another, so they are still very similiar Dockerfile/Dockerfile.gpu where the only difference is the install of tensorflow_gpu for the second one..

Pytorch is much better on that side, just takes whatever it finds :)

Anyway, I test both cpu/gpu version and it seems to work.

File: docker/Dockerfile.worker-gpu (hunk 5 above)
Comment (cthorey) [reply docker/Dockerfile.worker-gpu]: I usually like to have a passwd for security reason. Say, you spawn a worker onto the cloud, then without a passwd, anybody that figure out the IP of the instance can access the notebook. 

I get ride of it for the time being following @johannes69ki comment

File: docker/requirements.txt (hunk 6 above)
Comment (cthorey) [reply docker/requirements.txt]: Change the build context so that it can see those files

File: docker/jupyter_notebook_config.py (hunk 10 above)
Comment (cthorey) [reply docker/jupyter_notebook_config.py]: Tell me if I understand incorrectly, but i guess you can already do that. Let say you spin up the worker using **run_grid** in the cloud. Then you can access the notebook in you browser using IP_OF_YOUR_INSTANCE:8888

**Comment by cthorey** [comment]
Thx @ianlivingstone @johannes69ki for the comments, very helpful!! Especially leveraging the new version of nvidia-docker mask thing much easier. 

- I added a bash script to launch everything ( all 6 cases / gpu/no gpu and three grid modes). I add to modify a bit the ipfs_grid_worker_daemon.py to make it work - writing the whoaim file + overiding default args if GRID_MODE in the os.environ
- If someone can get the **runtime** config option working with compose, we can remove the uncessary complexity that arises from configure the docker daemon. For the moment, using the latest master (docker-compose version 1.20.0dev, build unknown), I got
 ```
ERROR: The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.nvidia: 'runtime'
```
- If someone is able to build/push the image to docker hub, we can then use them directly instead of building them using compose and modifying Dockerfile.gpu to build from openmined/grid

File: docker-compose.yml (hunk 2 above)
Comment (cthorey) [reply docker-compose.yml]: Actually, switch back to 2.3 to get the runtime option.

File: docker-compose.gpu.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (cthorey) [reply docker-compose.gpu.yml]: I finally get it working, it is only available in 2.3 version. https://docs.docker.com/compose/compose-file/compose-file-v2/#ports

**Comment by iamtrask** [comment]
Man this is an awesome portability improvement for the project. Please forgive the delay on my merge. That's on me. 

Installation and deployment for workers has been a HUGE issue - and this seems like it virtually solves it. Great work @cthorey and thank you for all the feedback @ianlivingstone and @johannes69ki !!
